adds table to support different context versions in the future

kdenhartog opened the pull request:
This coincides with PR https://github.com/w3c/did-core/pull/645 to help us avoid making normative errata changes when we want to update the context for did documents.


<!--
    This comment and the below content is programatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://pr-preview.s3.amazonaws.com/kdenhartog/did-core-registries/pull/251.html" title="Last updated on Mar 3, 2021, 1:41 AM UTC (c782c4a)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/did-spec-registries/251/bc9372e...kdenhartog:c782c4a.html" title="Last updated on Mar 3, 2021, 1:41 AM UTC (c782c4a)">Diff</a>

OR13 reviewed (approved):
This is great forward proofing... thanks @kdenhartog .... ping @dlongley @msporny

msporny commented on index.html:
```diff
@@ -1469,7 +1469,28 @@ <h4>@context</h4>
+        <table class="simple" style="width: 100%;">
+          <thead>
+            <tr>
+              <th>@Context Values</th>
```
```suggestion
              <th>@context Values</th>
```

msporny commented on index.html:
```diff
@@ -1469,7 +1469,28 @@ <h4>@context</h4>
+  String</a>.
+        </p>
+        <table class="simple" style="width: 100%;">
+          <thead>
```
We should have a link to the latest official version of the specification to go with each valid `@context` value.

kdenhartog replied:
Will add this

OR13 replied:
I am requesting changes on this, until notified this is done.

msporny reviewed (approved):
+1 to the addition as long as the proposed changes are included.

kdenhartog commented:
If we go with option 1 or 2 in this comment https://github.com/w3c/did-core/pull/645#discussion_r576453854 then I think it may be useful to separate out the context files from did-core and extensions/superset contexts. Additionally we should figure out the best way to version extension/contexts if we go with option 1 or 2 so feedback on #143. This is helpful so I can have a vague understanding of where everyone is at on versioning contexts and start making headway on that work. Right now I'm left to shoot blindly and hope I find something that satisfies others without any responses about how we want to handle it.

peacekeeper commented on index.html:
```diff
@@ -1469,7 +1469,28 @@ <h4>@context</h4>
 
-
+        <p>
+The following values are acceptable values for <code>@context</code> property as
```
```suggestion
The following values are acceptable values for the <code>@context</code> entry as
```

rhiaro commented:
> If this is the case, then we should only be including terms defined within the did-core spec inside the did context

@kdenhartog - I agree! I have raised this discrepancy multiple times. The normal JSON-LD behaviour would be for the core terms to be in the main context, and then implementers can pick and choose the extensions they need, and layer the contexts on top. Implementers may choose to import all of the contexts they need (after looking through the registry to see what is available) into a single additional one, for readability and ease of use.

I have accepted the argument that there are some terms - namely the LD suites defined in the security vocab - that are in common enough use that including them in the DID Core context is a reasonable convenience, but that defining them in the DID Core spec is out of scope. This would be an extra special list of exceptions which is frozen when the WG ends, which I have tried to describe in [the README](https://github.com/w3c/did-spec-registries/blob/master/vocabs/README.md). However since I wrote that (and thought it had been more or less agreed as a way forward) [another, more niche, term ended up in the core context too](https://github.com/w3c/did-spec-registries/issues/167#issuecomment-757490859).

I don't agree that we (the WG or Registries maintainers) should be attempting to provide extensions contexts, or putting third party extensions en masse into the Core. Implementations _need_ the freedom to pick and choose and to be intentional about what vocabs they are pulling in, rather than being served [a kitchen sink (see WG resolution about not doing this from May last year)](https://www.w3.org/2019/did-wg/Meetings/Minutes/2020-05-12-did-topic#resolution3) that they didn't necessarily bother to read or understand by default.

Since providing a JSON-LD context for all extension properties is a mandatory part of the registration process, the pieces are already in place.

If contexts that group multiple extensions together become available, these should be separate community resources rather than something that has to be maintained here, and hopefully would be tailored to particular use cases rather than catch-alls.

kdenhartog commented:
>I have accepted the argument that there are some terms - namely the LD suites defined in the security vocab - that are in common enough use that including them in the DID Core context is a reasonable convenience, but that defining them in the DID Core spec is out of scope. 

These are the main ones I want removed. They're the same ones causing us issues on updating the security vocab when also using the VC context and I highly suspect including these in the did-core context will just introduce more protected term conflicts in the future. 

I must have missed when this was originally discussed, or not had enough implementation experience to weigh in at the time, but I'm convinced more and more that catch all contexts are bad and only further produce implementation complexity for those who don't want to dive into the rabbit hole of figuring out how the json-ld tooling works. I think it would be good for us to properly separate these now and instead start setting the path that these suites and context definitions will fall under the purview of the LDP work in the CCG and WG in the future as many of us familiar with the LDP work would expect.

Had we opted to do the LDP work ahead of the VC/DID WGs this wouldn't have been a problem, but because of the order of which we've chosen to build these specs we've introduced a lot of headaches that are going to remain until we either get that work finished (and go back and clean up the issues we left in the DID/VC specs) or start properly separating the contexts to align with where the semantics are being defined.

Today the only solution I'm left with is to perform context overriding locally to prevent errors which is definitely going to cause interoperability concerns as we move into higher layers. Instead if we could fix the published contexts available remotely (because remote context inevitably get used when people don't understand how to properly setup their documentLoaders) then this would go a long way to solving these issues and helping to set good precedents as we move forward with our work.

kdenhartog commented:
Reading through some of my old responses I think my position on this has changed a bit since when I first raised this PR. At this time, the 4 main issues I'm looking to solve are:

1. Prevent `@protected` term conflicts between commonly used contexts listed in JSON-LD documents
2. Set good practices for how a json-ld context should be built
3. Future proof our work to prevent spec churn caused by json-ld related usage
4. Make it easier for developers who are being introduced to json-ld from VCs/DIDs to start working with it and developing vocabularies

With those in mind, I think problem 1 is the only one that needs to be addressed by the did-core WG so that only properties defined by the spec are included (I'm a bit worried about the verification method definitions being in did-core rather than LDP work, but we'll figure that out later).

The others can be handled elsewhere (TBD).

iherman commented:
I fundamentally agree with @rhiaro (https://github.com/w3c/did-spec-registries/pull/251#issuecomment-784527779). The current situation is all the more awkward that, at some point, we will (hopefully!) start a separate Working Group on Linked Data Canonicalization and Signature and the corresponding LDS vocabulary must then be controlled by _that_ Working Group. Keeping those terms in the DID context seems to tie that WG's hands, and that is not healthy.

In an ideal world, therefore, we should have several context files:

- DID core terms as defined by the spec and nothing else
- A separate LDS context (which, eventually, will be bound to a proper LDS vocabulary to be created, reproducing the work we have done for the DID vocabulary)
- Maybe a third context file would include the additional terms (like extra verification method types) that are formally registered. (The registry document is full of terms that are set to be defined in DID Core in terms of JSON-LD, which is utterly misleading and, I would say, wrong. See, e.g. [blockchainAccountId](https://w3c.github.io/did-spec-registries/#blockchainaccountid))

The question is: how close are we to an ideal world? How many current implementations would have problems, and is this acceptable at this point (we are not yet in CR after all!). I do not have an answer...

OR13 reviewed (changes requested):
Unclear if consensus is reached on the changes in this PR... please keep discussion on issues, and make it clear if you are for or against proposed changes. Its hard to tell what people want when discussing things other than the changeset

OR13 commented:
@iherman @rhiaro @kdenhartog can you each comment with what changes need to be made _to this pr_ before I can accept these changes, and open any issues to carry out related discussions elsewhere?

kdenhartog commented:
@OR13 I'm not sure this PR is going to have any changes based on tangent discussion about how contexts should be built. I've addressed all the feedback from @msporny and @peacekeeper and believe it should be ready to go.

I'll see if I can dig up the discussion where @rhiaro discussed this tagental context issue (where we include LDP suites in did-core context) and move the discussion back there. If not I'll open a new issue to discuss how to handle it.

TallTed commented on index.html:
```diff
@@ -1469,7 +1469,31 @@ <h4>@context</h4>
+        <table class="simple" style="width: 100%;">
+          <thead>
+            <tr>
+              <th>@Context Values</th>
```
```suggestion
              <th>`@context` Values</th>
```
The `C` is never capitalized in use. It's not the first actual character, so it's debatable whether it should ever be capitalized in title case. Here, I think it's pretty certain that it should *NOT* be capitalized.

OR13 replied:
No need for the ` since this is not markdown.

TallTed replied:
@OR13 - As I was recently informed by @msporny, Respec turns the backticks into `<code>` wrappers. I prefer the `<code>` tags myself, but I'm going along with the Editor's expressed preference for fewer chars in the spec source.

OR13 replied:
hmm, today I learned!

msporny replied:
Yeah, ReSpec understands Markdown. :)

kdenhartog replied:
TIL as well. thanks!

OR13 commented on index.html:
```diff
@@ -1469,7 +1469,31 @@ <h4>@context</h4>
+        <table class="simple" style="width: 100%;">
+          <thead>
+            <tr>
+              <th>@Context Values</th>
```
```suggestion
              <th>@context Values</th>
```
